swarm/chunk/tags.go

```diff
@@ -0,0 +1,211 @@
+package chunk
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math/rand"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+var (
+	errExists = errors.New("already exists")
+	errNoETA  = errors.New("unable to calculate ETA")
+)
+
+// tags holds the tag infos indexed by name
+type tags struct {
+	tags *sync.Map
+	rng  *rand.Rand
+}
+
+// NewTags creates a tags object
+func newTags() *tags {
+
+	return &tags{
+		tags: &sync.Map{},
+		rng:  rand.New(rand.NewSource(time.Now().Unix())),
+	}
+}
+
+// New creates a new tag, stores it by the name and returns it
+// it returns an error if the tag with this name already exists
+func (ts *tags) New(s string, total int) (*Tag, error) {
```

Comment:
could we move `Tag` struct declaration on top? maybe even move tags in a separate file?